samples/nrf9160/modem_shell/src/ppp/ppp_ctrl.c

```diff
@@ -0,0 +1,233 @@
+				     ppp_ctrl_net_mgmt_event_handler,
+				     (NET_EVENT_PPP_CARRIER_ON |
+				      NET_EVENT_PPP_CARRIER_OFF));
+
+	net_mgmt_init_event_callback(&ip_level_net_mgmt_event_cb,
+				     ppp_ctrl_net_mgmt_event_ip_levelhandler,
+				     (NET_EVENT_IPV4_ADDR_ADD |
+				      NET_EVENT_IPV4_ADDR_DEL));
+
+	net_mgmt_add_event_callback(&ppp_ctrl_net_mgmt_event_ppp_cb);
+	net_mgmt_add_event_callback(&ip_level_net_mgmt_event_cb);
+}
+
+/******************************************************************************/
+
+void ppp_ctrl_init(void)
+{
+	ppp_modem_data_socket_fd = PPP_MODEM_DATA_RAW_SCKT_FD_NONE;
+	ppp_data_socket_fd = PPP_MODEM_DATA_RAW_SCKT_FD_NONE;
+	ppp_ctrl_net_mgmt_events_subscribe();
+	ppp_iface_global = NULL;
+}
+
+/******************************************************************************/
+
+int ppp_ctrl_start(const struct shell *shell)
+{
+	struct ppp_context *ctx;
+	struct net_if *iface;
+	struct pdp_context_info_t *pdp_context_info;
+
+	ppp_shell_global = shell;
+	int idx = 0; /* Note: PPP iface index assumed to be 0 */
+
+	if (ppp_modem_data_socket_fd != PPP_MODEM_DATA_RAW_SCKT_FD_NONE) {
+		shell_warn(shell, "PPP already up.\n");
+		goto return_error;
+	}
+
+	ctx = net_ppp_context_get(idx);
+	if (!ctx) {
+		shell_error(shell, "PPP context not found.\n");
+		goto return_error;
+	}
+
+	/* Note: no multicontext support, only at default PDN context */
+	pdp_context_info = ltelc_api_get_pdp_context_info_by_pdn_cid(0);
+	if (pdp_context_info == NULL) {
+		shell_error(shell, "PPP context not found.\n");
+		goto return_error;
+	}
+
+	iface = ctx->iface;
+	ppp_iface_global = iface;
+
+	net_if_flag_set(iface, NET_IF_NO_AUTO_START);
+
+	memcpy(&(ctx->ipcp.my_options.address), &(pdp_context_info->ip_addr4),
+	       sizeof(ctx->ipcp.my_options.address));
+	memcpy(&ctx->ipcp.my_options.dns1_address,
+	       &pdp_context_info->dns_addr4_primary,
+	       sizeof(ctx->ipcp.my_options.dns1_address));
+	memcpy(&ctx->ipcp.my_options.dns2_address,
+	       &pdp_context_info->dns_addr4_secondary,
+	       sizeof(ctx->ipcp.my_options.dns2_address));
+
+	free(pdp_context_info);
+
+	ppp_modem_data_socket_fd = socket(AF_PACKET, SOCK_RAW, 0);
+	if (ppp_modem_data_socket_fd < 0) {
+		shell_error(shell,
+			    "modem data socket creation failed: (%d)!!!!\n",
+			    -errno);
+		goto return_error;
+	} else {
+		shell_info(shell, "modem data socket %d created for modem data",
+			   ppp_modem_data_socket_fd);
+	}
+
+#ifdef SO_SNDTIMEO
+	struct timeval tv;
+
+	/* blocking socket and we do not want to block for long: 3 sec timeout for sending: */
+	tv.tv_sec = 3;
+	tv.tv_usec = 0;
+	if (setsockopt(ppp_modem_data_socket_fd, SOL_SOCKET, SO_SNDTIMEO,
+		       (struct timeval *)&tv, sizeof(struct timeval)) < 0) {
+		shell_error(shell, "Unable to set socket SO_SNDTIMEO");
+	}
+#endif
+
+	net_if_up(iface);
+
+	/* Create raw Zephyr socket for passing data to/from ppp link: */
+	ppp_data_socket_fd = socket(AF_PACKET, SOCK_RAW, IPPROTO_RAW);
```

Comment:
I'm actually surprised to see it works this way, probably only because packet socket from Zepyr is processed before the nRF91 offloading layer (which is only by luck, nothing guarantees that). Basically, the problem is that currently the viability of particular socket implementation for a specific famliy/type/protocol combo is determined by the `is_supported()` function registred with a socket. Given that such combo could be supported by both, native socket implementation and the offloaded one, this makes specific uncertainity on which implementation should Zephyr choose for a particular socket.

Reply:
@rlubos yes, it has been working. A very good point, but how to make this bullet proof? Shall we add stricter conditions in offloading socket is_supported() or make it sure that zephyr socket are always checked 1st? Anyways, I think that in this case, if compiled in, the zephyr packet socket should be the only owner for this combination.

Reply:
Added a condition to nrf91_socket is_supported() to check this case if zephyr packet socket is used. @rlubos , please check commit: https://github.com/trantanen/sdk-nrf/commit/81df6815e5ce7b172444ddb46902d849b96d7feb

Reply:
Looks ok to me.